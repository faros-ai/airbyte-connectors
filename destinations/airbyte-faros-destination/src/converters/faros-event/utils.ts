import {ErrorObject} from 'ajv';
import {groupBy} from 'lodash';

import {Artifact, FarosPath, FarosTag, RunStep} from './types';

interface OrganizationKey {
  readonly uid: string;
  readonly source: string;
}

export interface DeploymentKey {
  readonly uid: string;
  readonly source: string;
}

export interface DeployIds {
  readonly id: string;
  readonly source: string;
  readonly environment: string;
  readonly application: string;
}

export interface ArtifactKey {
  readonly uid: string;
  readonly repository: {
    readonly uid: string;
    readonly organization: OrganizationKey;
  };
}

export interface ArtifactIds {
  readonly id: string;
  readonly repository: string;
  readonly organization: string;
  readonly source: string;
}

export interface CommitKey {
  readonly sha: string;
  readonly repository: {
    readonly name: string;
    readonly organization: OrganizationKey;
  };
}

export interface CommitIds {
  readonly sha: string;
  readonly repository: string;
  readonly organization: string;
  readonly source: string;
}

export interface BuildKey {
  readonly uid: string;
  readonly pipeline: {
    readonly uid: string;
    readonly organization: OrganizationKey;
  };
}

export interface BuildIds {
  readonly id: string;
  readonly pipeline: string;
  readonly organization: string;
  readonly source: string;
}

export interface BuildStepKey {
  readonly uid: string;
  readonly build: BuildKey;
}

export interface AgentKey {
  readonly uid: string;
  readonly source: string;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveCommitIds(commit: any): CommitIds {
  if (commit.uri) {
    const parsedUri = parseUri(commit.uri);
    return {
      sha: parsedUri[3],
      repository: parsedUri[2].toLowerCase(),
      organization: parsedUri[1].toLowerCase(),
      source: parsedUri[0],
    };
  }
  return {
    sha: commit.sha,
    repository: commit.repository.toLowerCase(),
    organization: commit.organization.toLowerCase(),
    source: commit.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeCommitKey(commit: any): CommitKey {
  const commitIds = resolveCommitIds(commit);
  return {
    sha: commitIds.sha,
    repository: {
      name: commitIds.repository,
      organization: {
        uid: commitIds.organization,
        source: commitIds.source,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveBuildIds(run: any): BuildIds {
  if (run.uri) {
    const parsedUri = parseUri(run.uri);
    return {
      id: parsedUri[3],
      pipeline: parsedUri[2],
      organization: parsedUri[1],
      source: parsedUri[0],
    };
  }
  return {
    id: run.id,
    pipeline: run.pipeline,
    organization: run.organization,
    source: run.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeBuildKey(run: any): BuildKey {
  const buildIds = resolveBuildIds(run);
  return {
    uid: buildIds.id,
    pipeline: {
      uid: buildIds.pipeline,
      organization: {
        uid: buildIds.organization,
        source: buildIds.source,
      },
    },
  };
}

export function makeBuildStepKey(step: RunStep, build: BuildKey): BuildStepKey {
  return {
    uid: step.id,
    build,
  };
}

export function resolveDeploymentIds(deploy: any): DeployIds {
  let deployIds;
  if (deploy.uri) {
    const parsedUri = parseUri(deploy.uri);
    deployIds = {
      id: parsedUri[3],
      environment: parsedUri[2],
      application: parsedUri[1],
      source: parsedUri[0],
    };
  } else {
    deployIds = {
      id: deploy.id,
      environment: deploy.environment,
      application: deploy.application,
      source: deploy.source,
    };
  }
  return deployIds;
}

function enumToMap(enumValues: string[]): Map<string, string> {
  return new Map(enumValues.map((value) => [value.toLowerCase(), value]));
}

function resolveEnum(
  enumValue: Map<string, string>,
  field: string,
  fieldDetail?: string | null
): {category: string; detail?: string | null} {
  const lowerField = field.toLowerCase();
  const value = enumValue.get(lowerField);
  if (value) {
    return {category: value, detail: fieldDetail};
  }
  return {
    category: 'Custom',
    detail: field.concat(fieldDetail ? ` - ${fieldDetail}` : ''),
  };
}

const deploymentEnvironments = enumToMap([
  'Prod',
  'Staging',
  'QA',
  'Dev',
  'Sandbox',
  'Canary',
  'Custom',
]);

export function resolveDeploymentEnvironment(
  environment: string,
  environmentDetails?: string | null
): {category: string; detail?: string | null} {
  return resolveEnum(deploymentEnvironments, environment, environmentDetails);
}

const deploymentStatuses = enumToMap([
  'Canceled',
  'Custom',
  'Failed',
  'Queued',
  'Running',
  'RolledBack',
  'Success',
]);

export function resolveDeploymentStatus(
  status?: string,
  statusDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!status) {
    return undefined;
  }
  return resolveEnum(deploymentStatuses, status, statusDetails);
}

const runStatuses = enumToMap([
  'Canceled',
  'Custom',
  'Failed',
  'Queued',
  'Running',
  'Success',
  'Unknown',
]);

export function resolveRunStatus(
  status?: string,
  statusDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!status) {
    return undefined;
  }
  return resolveEnum(runStatuses, status, statusDetails);
}

const runStepTypes = enumToMap(['Custom', 'Manual', 'Script']);

export function resolveRunStepType(
  type?: string,
  typeDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!type) {
    return undefined;
  }
  return resolveEnum(runStepTypes, type, typeDetails);
}

const computeAgentOS = enumToMap(['Linux', 'MacOS', 'Windows', 'Custom']);

export function resolveComputeAgentOS(
  os?: string,
  osDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!os) {
    return undefined;
  }
  return resolveEnum(computeAgentOS, os, osDetails);
}

const metricDefinitionTypes = enumToMap([
  'Numeric',
  'String',
  'Date',
  'Timestamp',
  'Duration',
  'Percentage',
  'Custom',
]);

export function resolveMetricDefinitionType(
  type?: string,
  typeDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!type) {
    return undefined;
  }
  return resolveEnum(metricDefinitionTypes, type, typeDetails);
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveArtifactIds(artifact: any): ArtifactIds {
  if (artifact.uri) {
    const parsedUri = parseUri(artifact.uri);
    return {
      id: parsedUri[3],
      repository: parsedUri[2].toLowerCase(),
      organization: parsedUri[1].toLowerCase(),
      source: parsedUri[0],
    };
  }
  return {
    id: artifact.id,
    repository: artifact.repository.toLowerCase(),
    organization: artifact.organization.toLowerCase(),
    source: artifact.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeArtifactKey(artifact: any): ArtifactKey {
  const artifactIds = resolveArtifactIds(artifact);
  return {
    uid: artifactIds.id,
    repository: {
      uid: artifactIds.repository.toLowerCase(),
      organization: {
        uid: artifactIds.organization.toLowerCase(),
        source: artifactIds.source,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeDummyArtifact(commit: any): Artifact {
  const commitKey = makeCommitKey(commit);
  return {
    id: commitKey.sha,
    repository: commitKey.repository.name,
    organization: commitKey.repository.organization.uid,
    source: commitKey.repository.organization.source,
  };
}

// Convert ISO-8601 string, number string, number, or 'Now' into milliseconds
export function getTime(date?: string | number): number | undefined {
  if (date) {
    if (typeof date === 'string') {
      if (date === 'Now') {
        return Date.now();
      }
      const dateInt = Number(date);
      if (!isNaN(dateInt)) {
        return new Date(dateInt).getTime();
      }
    }
    return new Date(date).getTime();
  }
  return undefined;
}

// Convert ISO-8601 string, number string, number, or 'Now' into ISO-8601 string
export function toDate(date?: string | number): string | undefined {
  if (date) {
    if (typeof date === 'string') {
      if (date === 'Now') {
        return new Date(Date.now()).toISOString();
      }
      const dateInt = Number(date);
      if (!isNaN(dateInt)) {
        return new Date(dateInt).toISOString();
      }
    }
    return new Date(date).toISOString();
  }
  return undefined;
}

export function parseTags(tags?: FarosTag[] | string): FarosTag[] | undefined {
  const parser = (input: string): FarosTag[] => {
    const resTags: FarosTag[] = [];
    const keyValues = input.split(',');
    for (const keyValue of keyValues) {
      const splitTag = keyValue.split(':');
      resTags.push({key: splitTag[0], value: splitTag[1]});
    }
    return resTags;
  };
  return parseArray(parser, tags);
}

export function parsePaths(
  paths?: FarosPath[] | string
): FarosPath[] | undefined {
  const parser = (input: string): FarosPath[] => {
    return input.split(',').map<FarosPath>((path) => {
      return {path};
    });
  };
  return parseArray(parser, paths);
}

export function parsePullRequestNumber(
  number: string | number
): number | undefined {
  const prNum = typeof number === 'string' ? Number(number) : number;
  return isNaN(prNum) ? undefined : prNum;
}

export function parseUri(uri: string): string[] {
  const components: string[] = [];
  const split1 = uri.split('://');
  components.push(split1[0]);
  split1[1].split('/').forEach((value) => components.push(value));
  return components;
}

export function parseStringArray(
  array?: string[] | string,
  separator = ','
): string[] | undefined {
  return parseArray((s) => s.split(separator), array);
}

function parseArray<T>(
  parser: (input: string) => T[],
  input?: T[] | string
): T[] | undefined {
  if (typeof input === 'string') {
    return parser(input);
  }
  return input;
}

export interface EventValidationError {
  readonly message: string;
  readonly details?: ReadonlyArray<string>;
  readonly allowedValues?: ReadonlyArray<string>;
  readonly additionalProperties?: ReadonlyArray<string>;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function removeUndefinedProperties(object: any): any {
  const result = {...object};
  Object.keys(result).forEach(
    (key) => result[key] === undefined && delete result[key]
  );
  return result;
}

export function formatErrors(
  errors?: ErrorObject[] | null,
  rootObject?: string
): ReadonlyArray<EventValidationError> {
  if (!errors) {
    return [];
  }

  const formattedErrors: EventValidationError[] = [];
  const groupedErrors = groupBy<ErrorObject>(errors, (err) => {
    return formatErrorDetailsMessage(err, rootObject);
  });

  for (const [message, errs] of Object.entries(groupedErrors)) {
    const details = errs.flatMap(
      (err) =>
        err.params?.errors?.map((v: any) =>
          formatErrorDetailsMessage(v, rootObject)
        ) ?? []
    );
    const additionalProperties = errs.flatMap(
      (err) => err.params?.additionalProperty ?? []
    );
    formattedErrors.push({
      message,
      details: details.length > 0 ? details : undefined,
      allowedValues: getAllowedValues(errs),
      additionalProperties:
        additionalProperties.length > 0 ? additionalProperties : undefined,
    });
  }
  return formattedErrors;
}

function formatErrorDetailsMessage(
  error: ErrorObject,
  rootObject = 'data'
): string {
  const prefix = error?.instancePath
    ? `${rootObject}${error.instancePath.replace(/\//g, '.')} `
    : `${rootObject} `;
  return `${prefix}${error.message}`;
}

function getAllowedValues(errors: ErrorObject[]): string[] | undefined {
  let result;
  if (errors?.length > 0) {
    result = errors[0].params?.allowedValues;
    if (!result && errors[0].params?.errors?.length > 0) {
      result = errors[0].params?.errors[0].params?.allowedValues;
    }
  }
  return result;
}

// Max length for free-form description text fields such as issue body
const MAX_TEXT_LENGTH = 1000;
export function cleanAndTruncate(
  str?: string,
  maxLength?: number
): string | null | undefined {
  if (!str) {
    return str;
  }
  const length = maxLength ?? MAX_TEXT_LENGTH;
  let result: string;
  if (str.length <= length) {
    result = str;
  } else {
    // If the last character is part of a unicode surrogate pair, include the next character
    const lastChar = str.codePointAt(length - 1) ?? 0;
    result =
      lastChar > 65535
        ? str.substring(0, length + 1)
        : str.substring(0, length);
  }
  // eslint-disable-next-line no-control-regex
  return result.replace(/\u0000/g, '');
}
