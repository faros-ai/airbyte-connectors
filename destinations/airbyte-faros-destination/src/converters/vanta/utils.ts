import fs from 'fs-extra';
import path from 'path';

// Function to find a specific parent directory
function findParentDirectory(
  currentDir: string,
  targetDirName: string
): string | null {
  let currentPath = currentDir;
  while (currentPath !== '/') {
    if (path.basename(currentPath) === targetDirName) {
      return currentPath;
    }
    currentPath = path.dirname(currentPath);
  }
  return null;
}

export function getQueryFromName(name: string): string {
  const crt_dir = __dirname;
  // We go back to the part of dir which ends with name:
  const targetDirPath = findParentDirectory(
    crt_dir,
    'airbyte-faros-destination'
  );
  if (!targetDirPath) {
    throw new Error('airbyte-faros-destination directory not found');
  }

  const gql_dir = path.join(targetDirPath, 'resources');

  const fn = `${name}.gql`;
  // Check if fn is in gql_dir:
  if (!fs.existsSync(`${gql_dir}/${fn}`)) {
    throw new Error(`Query ${fn} not found in ${gql_dir}`);
  }
  // We access the directory resources/gql_queries to get the query
  return fs.readFileSync(`${gql_dir}/${fn}`, 'utf8');
}

export function looksLikeGitCommitSha(sha: string): boolean {
  return /^[a-f0-9]{40}$/i.test(sha);
}
