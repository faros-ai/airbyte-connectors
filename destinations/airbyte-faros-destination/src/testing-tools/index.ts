export * from './cli';
export * from './testing-tools';
export * from './utils';
