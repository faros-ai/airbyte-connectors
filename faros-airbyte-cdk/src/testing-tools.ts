import fs from 'fs';
import {Dictionary} from 'ts-essentials';

import {
  AirbyteConfig,
  AirbyteConfiguredCatalog,
  AirbyteMessageType,
  AirbyteRecord,
  AirbyteSourceBase,
  AirbyteState,
} from '.';

export function readTestResourceFile(fileName: string): string {
  return fs.readFileSync(`test/resources/${fileName}`, 'utf8');
}

/**
 * Parse a test resource into JSON
 */
export function readTestResourceAsJSON(fileName: string): any {
  return JSON.parse(readTestResourceFile(fileName));
}

export interface SourceCheckTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath?: string | AirbyteConfig;
}

export const sourceCheckTest = async (
  options: SourceCheckTestOptions
): Promise<void> => {
  const {source, configOrPath} = options;
  const config = resolveInput(configOrPath);
  expect(await source.check(config)).toMatchSnapshot();
};

export interface SourceReadTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath: string | AirbyteConfig;
  catalogOrPath: string | AirbyteConfiguredCatalog;
  stateOrPath?: string | AirbyteState;
  onBeforeReadResultConsumer?: (res: {
    config: AirbyteConfig;
    catalog: AirbyteConfiguredCatalog;
    state?: AirbyteState;
  }) => void;
  checkRecordsData?: (records: ReadonlyArray<Dictionary<any>>) => void;
}

export const sourceReadTest = async (
  options: SourceReadTestOptions
): Promise<void> => {
  const {
    source,
    configOrPath,
    catalogOrPath,
    stateOrPath = undefined,
    onBeforeReadResultConsumer = undefined,
    checkRecordsData = undefined,
  } = options;
  const config = resolveInput(configOrPath);
  const catalog = resolveInput(catalogOrPath);
  const state = stateOrPath ? resolveInput(stateOrPath) : {};
  const res = await source.onBeforeRead(config, catalog, state);
  if (onBeforeReadResultConsumer) {
    onBeforeReadResultConsumer(res);
  }
  const generator = source.read(res.config, res.config, res.catalog, res.state);
  const records: Dictionary<any>[] = [];
  for await (const message of generator) {
    if (message.type === AirbyteMessageType.RECORD) {
      records.push((message as AirbyteRecord).record.data);
    }
  }
  if (checkRecordsData) {
    checkRecordsData(records);
  }
};

function resolveInput<T>(inputOrPath: string | T): T {
  return typeof inputOrPath === 'string'
    ? readTestResourceAsJSON(inputOrPath)
    : inputOrPath;
}
