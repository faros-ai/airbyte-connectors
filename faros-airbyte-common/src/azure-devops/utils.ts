import {User} from './types';

/**
 * Gets a unique identifier for an Azure DevOps user.
 * Tries principalName first, then falls back to uniqueName.
 * This handles differences between Azure DevOps Cloud and Server.
 *
 * @param user - The Azure DevOps user object
 * @returns The user's unique identifier, or undefined if neither field exists
 */
export function getUserIdentifier(user: User): string | undefined {
  if ('principalName' in user && user.principalName) {
    return user.principalName;
  }
  if ('uniqueName' in user && user.uniqueName) {
    return user.uniqueName;
  }
  return undefined;
}
