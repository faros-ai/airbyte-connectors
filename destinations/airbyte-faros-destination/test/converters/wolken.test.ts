import {getLocal} from 'mockttp';

import {initMockttp, tempConfig} from '../testing-tools';
import {destinationWriteTest} from './utils';

describe('wolken', () => {
  const mockttp = getLocal({debug: false, recordTraffic: false});
  let configPath: string;

  beforeEach(async () => {
    await initMockttp(mockttp);
    configPath = await tempConfig({
      api_url: mockttp.url,
      log_records: true,
    });
  });

  afterEach(async () => {
    await mockttp.stop();
  });

  describe('users', () => {
    test('users', async () => {
      await destinationWriteTest({
        configPath,
        catalogPath: 'test/resources/wolken/catalog.json',
        inputRecordsPath: 'wolken/users.log',
        checkRecordsData: (records) => expect(records).toMatchSnapshot(),
      });
    });
  });

  describe('incidents', () => {
    test('incidents', async () => {
      await destinationWriteTest({
        configPath,
        catalogPath: 'test/resources/wolken/catalog.json',
        inputRecordsPath: 'wolken/incidents.log',
        checkRecordsData: (records) => expect(records).toMatchSnapshot(),
      });
    });

    test('incident application impact', async () => {
      configPath = await tempConfig({
        api_url: mockttp.url,
        log_records: true,
        source_specific_configs: {
          wolken: {
            service_id_flex_field_name: 'Service ID',
            application_mapping: {
              A3F91B6D: {name: 'Test App'},
            },
            store_current_incidents_associations: true,
          },
        },
      });
      await destinationWriteTest({
        configPath,
        catalogPath: 'test/resources/wolken/catalog.json',
        inputRecordsPath: 'wolken/incidents_configuration_items.log',
        checkRecordsData: (records) => expect(records).toMatchSnapshot(),
      });
    });
  });

  describe('configuration_items', () => {
    test('configuration_items', async () => {
      configPath = await tempConfig({
        api_url: mockttp.url,
        log_records: true,
        source_specific_configs: {
          wolken: {
            service_id_flex_field_name: 'Service ID',
            jira_project_key_flex_field_name: 'JIRA Key',
            application_tag_flex_field_names: ['Environment'],
            application_tag_flex_field_user_lookup_names: ['Engineering Owner'],
            project_tag_flex_field_names: ['Engineering Owner'],
            path_hierarchy_flex_field_names: ['Service ID'],
            application_mapping: {
              A3F91B6D: {name: 'Test App'},
            },
            user_lookup_extra_fields_mapping: {
              Department: 'userAddress.departmentName',
            },
          },
        },
      });
      await destinationWriteTest({
        configPath,
        catalogPath: 'test/resources/wolken/catalog.json',
        inputRecordsPath: 'wolken/configuration_items.log',
        checkRecordsData: (records) => expect(records).toMatchSnapshot(),
      });
    });
  });
});
