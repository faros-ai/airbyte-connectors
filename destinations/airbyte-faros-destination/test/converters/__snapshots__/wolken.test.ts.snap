// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wolken configuration_items configuration_items 1`] = `
Array [
  "Processed 2 records",
  "Processed records by stream: {\\\\\\"mytestsource__wolken__configuration_items\\\\\\":1,\\\\\\"mytestsource__wolken__users\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 13 records",
  "Would write records by model: {\\\\\\"compute_Application\\\\\\":1,\\\\\\"compute_ApplicationPath\\\\\\":1,\\\\\\"compute_ApplicationTag\\\\\\":3,\\\\\\"faros_Path\\\\\\":1,\\\\\\"faros_Tag\\\\\\":4,\\\\\\"ims_User\\\\\\":1,\\\\\\"tms_ProjectPath\\\\\\":1,\\\\\\"tms_ProjectTag\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;

exports[`wolken configuration_items configuration_items 2`] = `
Array [
  Object {
    "ims_User": Object {
      "email": "john.doe@example.com",
      "name": "John Smith",
      "source": "Wolken",
      "uid": "12345",
    },
  },
  Object {
    "compute_Application": Object {
      "name": "Test App",
      "platform": "unknown",
      "source": "Wolken",
      "uid": "Test App_unknown",
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Environment",
      "source": "Wolken",
      "uid": "Environment__Production",
      "value": "Production",
    },
  },
  Object {
    "compute_ApplicationTag": Object {
      "application": Object {
        "name": "Test App",
        "platform": "unknown",
        "source": "Wolken",
        "uid": "Test App_unknown",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Environment__Production",
      },
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Engineering Owner",
      "source": "Wolken",
      "uid": "Engineering Owner__john.doe@example.com",
      "value": "john.doe@example.com",
    },
  },
  Object {
    "compute_ApplicationTag": Object {
      "application": Object {
        "name": "Test App",
        "platform": "unknown",
        "source": "Wolken",
        "uid": "Test App_unknown",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Engineering Owner__john.doe@example.com",
      },
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Engineering Owner Department",
      "source": "Wolken",
      "uid": "Engineering Owner Department__DEP1",
      "value": "DEP1",
    },
  },
  Object {
    "compute_ApplicationTag": Object {
      "application": Object {
        "name": "Test App",
        "platform": "unknown",
        "source": "Wolken",
        "uid": "Test App_unknown",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Engineering Owner Department__DEP1",
      },
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Engineering Owner",
      "source": "Wolken",
      "uid": "Engineering Owner__john.doe@example.com",
      "value": "john.doe@example.com",
    },
  },
  Object {
    "tms_ProjectTag": Object {
      "project": Object {
        "source": "Jira",
        "uid": "FAROS",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Engineering Owner__john.doe@example.com",
      },
    },
  },
  Object {
    "faros_Path": Object {
      "path": "A3F91B6D",
      "source": "Wolken",
      "uid": "A3F91B6D",
    },
  },
  Object {
    "compute_ApplicationPath": Object {
      "application": Object {
        "name": "Test App",
        "platform": "unknown",
        "source": "Wolken",
        "uid": "Test App_unknown",
      },
      "path": Object {
        "uid": "A3F91B6D",
      },
      "source": "Wolken",
    },
  },
  Object {
    "tms_ProjectPath": Object {
      "path": Object {
        "uid": "A3F91B6D",
      },
      "project": Object {
        "source": "Jira",
        "uid": "FAROS",
      },
      "source": "Wolken",
    },
  },
]
`;

exports[`wolken incidents incident application impact 1`] = `
Array [
  "Processed 3 records",
  "Processed records by stream: {\\\\\\"mytestsource__wolken__configuration_items\\\\\\":1,\\\\\\"mytestsource__wolken__incidents\\\\\\":2}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 16 records",
  "Would write records by model: {\\\\\\"__Flush\\\\\\":1,\\\\\\"compute_Application\\\\\\":1,\\\\\\"faros_Tag\\\\\\":4,\\\\\\"ims_Incident\\\\\\":2,\\\\\\"ims_IncidentApplicationImpact\\\\\\":1,\\\\\\"ims_IncidentApplicationImpact__Deletion\\\\\\":2,\\\\\\"ims_IncidentAssignment\\\\\\":1,\\\\\\"ims_IncidentFarosTag\\\\\\":4}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;

exports[`wolken incidents incident application impact 2`] = `
Array [
  Object {
    "ims_Incident": Object {
      "createdAt": "2025-01-08T09:54:48.000Z",
      "description": "Support notification incident requiring investigation",
      "priority": Object {
        "category": "Low",
        "detail": "Low - P4",
      },
      "resolvedAt": "2025-01-23T09:02:12.000Z",
      "severity": Object {
        "category": "Sev3",
        "detail": "Medium",
      },
      "source": "Wolken",
      "status": Object {
        "category": "Resolved",
        "detail": "Closed",
      },
      "title": "Inc_Support_Notification",
      "uid": "123456",
      "updatedAt": "2025-01-23T09:02:39.000Z",
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Category",
      "source": "Wolken",
      "uid": "Category__N/A",
      "value": "N/A",
    },
  },
  Object {
    "ims_IncidentFarosTag": Object {
      "incident": Object {
        "source": "Wolken",
        "uid": "123456",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Category__N/A",
      },
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Subcategory",
      "source": "Wolken",
      "uid": "Subcategory__N/A",
      "value": "N/A",
    },
  },
  Object {
    "ims_IncidentFarosTag": Object {
      "incident": Object {
        "source": "Wolken",
        "uid": "123456",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Subcategory__N/A",
      },
    },
  },
  Object {
    "ims_IncidentAssignment": Object {
      "assignee": Object {
        "source": "Wolken",
        "uid": "10002",
      },
      "incident": Object {
        "source": "Wolken",
        "uid": "123456",
      },
      "source": "Wolken",
    },
  },
  Object {
    "compute_Application": Object {
      "name": "Test App",
      "platform": "unknown",
      "source": "Wolken",
      "uid": "Test App_unknown",
    },
  },
  Object {
    "ims_Incident": Object {
      "createdAt": "2025-01-08T09:54:48.000Z",
      "description": "Support notification incident requiring investigation",
      "priority": Object {
        "category": "Low",
        "detail": "Low - P4",
      },
      "resolvedAt": "2025-01-23T09:02:12.000Z",
      "severity": Object {
        "category": "Sev3",
        "detail": "Medium",
      },
      "source": "Wolken",
      "status": Object {
        "category": "Resolved",
        "detail": "Closed",
      },
      "title": "Inc_Support_Notification",
      "uid": "123457",
      "updatedAt": "2025-01-23T09:02:39.000Z",
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Category",
      "source": "Wolken",
      "uid": "Category__N/A",
      "value": "N/A",
    },
  },
  Object {
    "ims_IncidentFarosTag": Object {
      "incident": Object {
        "source": "Wolken",
        "uid": "123457",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Category__N/A",
      },
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Subcategory",
      "source": "Wolken",
      "uid": "Subcategory__N/A",
      "value": "N/A",
    },
  },
  Object {
    "ims_IncidentFarosTag": Object {
      "incident": Object {
        "source": "Wolken",
        "uid": "123457",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Subcategory__N/A",
      },
    },
  },
  Object {
    "ims_IncidentApplicationImpact__Deletion": Object {
      "flushRequired": false,
      "source": "Wolken",
      "where": Object {
        "incident": Object {
          "source": "Wolken",
          "uid": "123456",
        },
      },
    },
  },
  Object {
    "ims_IncidentApplicationImpact__Deletion": Object {
      "flushRequired": false,
      "source": "Wolken",
      "where": Object {
        "incident": Object {
          "source": "Wolken",
          "uid": "123457",
        },
      },
    },
  },
  Object {
    "__Flush": Object {
      "source": "Wolken",
    },
  },
  Object {
    "ims_IncidentApplicationImpact": Object {
      "application": Object {
        "name": "Test App",
        "platform": "unknown",
        "uid": "Test App_unknown",
      },
      "incident": Object {
        "source": "Wolken",
        "uid": "123456",
      },
      "source": "Wolken",
    },
  },
]
`;

exports[`wolken incidents incidents 1`] = `
Array [
  "Processed 1 records",
  "Processed records by stream: {\\\\\\"mytestsource__wolken__incidents\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 6 records",
  "Would write records by model: {\\\\\\"faros_Tag\\\\\\":2,\\\\\\"ims_Incident\\\\\\":1,\\\\\\"ims_IncidentAssignment\\\\\\":1,\\\\\\"ims_IncidentFarosTag\\\\\\":2}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;

exports[`wolken incidents incidents 2`] = `
Array [
  Object {
    "ims_Incident": Object {
      "createdAt": "2025-01-08T09:54:48.000Z",
      "description": "Support notification incident requiring investigation",
      "priority": Object {
        "category": "Low",
        "detail": "Low - P4",
      },
      "resolvedAt": "2025-01-23T09:02:12.000Z",
      "severity": Object {
        "category": "Sev3",
        "detail": "Medium",
      },
      "source": "Wolken",
      "status": Object {
        "category": "Resolved",
        "detail": "Closed",
      },
      "title": "Inc_Support_Notification",
      "uid": "123456",
      "updatedAt": "2025-01-23T09:02:39.000Z",
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Category",
      "source": "Wolken",
      "uid": "Category__N/A",
      "value": "N/A",
    },
  },
  Object {
    "ims_IncidentFarosTag": Object {
      "incident": Object {
        "source": "Wolken",
        "uid": "123456",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Category__N/A",
      },
    },
  },
  Object {
    "faros_Tag": Object {
      "key": "Subcategory",
      "source": "Wolken",
      "uid": "Subcategory__N/A",
      "value": "N/A",
    },
  },
  Object {
    "ims_IncidentFarosTag": Object {
      "incident": Object {
        "source": "Wolken",
        "uid": "123456",
      },
      "source": "Wolken",
      "tag": Object {
        "uid": "Subcategory__N/A",
      },
    },
  },
  Object {
    "ims_IncidentAssignment": Object {
      "assignee": Object {
        "source": "Wolken",
        "uid": "10002",
      },
      "incident": Object {
        "source": "Wolken",
        "uid": "123456",
      },
      "source": "Wolken",
    },
  },
]
`;

exports[`wolken users users 1`] = `
Array [
  "Processed 1 records",
  "Processed records by stream: {\\\\\\"mytestsource__wolken__users\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 1 records",
  "Would write records by model: {\\\\\\"ims_User\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;

exports[`wolken users users 2`] = `
Array [
  Object {
    "ims_User": Object {
      "email": "test.user@company.com",
      "name": "John Smith",
      "source": "Wolken",
      "uid": "12345",
    },
  },
]
`;
